src/extension.ts

```diff
@@ -1,13 +1,25 @@
 "use strict";
 
 import * as vscode from "vscode";
+import { Environment } from "./environmentPath";
 import { init as initLocalize } from "./localize";
+import { InstanceManagerService } from "./service/instanceManager.service";
+import { state } from "./state";
 import { Sync } from "./sync";
 
 export async function activate(context: vscode.ExtensionContext) {
+  state.context = context;
+  state.environment = new Environment();
+
+  if (!InstanceManagerService.instanceSet()) {
+    InstanceManagerService.setInstance();
+  }
+
+  setInterval(InstanceManagerService.checkAndUpdate, 60000);

```

Comment:
why are we using `setInterval`. Is there anything better we can use?

Reply:
This is the only thing I could come up with to check if the original instance had been closed. Do you know a better way to handle this? I tried to use the `deactivate` function that VSCode provides to extensions, but it looks like that doesn't fire when the editor is closed.

Reply:
> I tried to use the deactivate function that VSCode provides to extensions

If you think we can achieve through `deactivate` function ( if it works well ) then we should definitely use it. You can ask vscode team about `deactivate` function if it doesnt perform.

Have you checked this :  https://github.com/Microsoft/vscode/issues/47881

Reply:
> The vscode API will be unreliable at deactivation time

I'm guessing that's why it didn't work because we have to use the API to reset the global state.

However, I think we can bypass using global state entirely. We can make it so only the focused window has auto upload priveleges. This would be a much cleaner solution instead of writing to global state all the time. What do you think?

Reply:
> We can make it so only the focused window has auto upload priveleges

Yes this would be good instead of writing. How you intend to do this ? What if user try upload and switch window and try upload from there, we need to consider all those scenarios.

Reply:
> What if user try upload and switch window and try upload from there

I think we can leave that up to the lockfile. If the first window is already uploading, the second one won't be able to trigger auto upload until the first one is done and unlocks the lockfile.

> How you intend to do this?

There is a property `vscode.window.state.focused` that we can use to check if the current window the extension is running is focused.

Reply:
Nice. lets go with it. It will be very much better comparing with writing and maintaining the states all the times.

Reply:
Great! I have already pushed a commit with those changes, so it should be ready for merging.